src/client/visualizations/heat-map/labeled-heatmap.tsx

```diff
@@ -0,0 +1,214 @@
+import { HeatmapHighlightModal } from "./heatmap-highlight-modal";
+import { HeatmapHoverIndicator } from "./heatmap-hover-indicator";
+import { HeatmapHoverTooltip } from "./heatmap-hover-tooltip";
+import { HeatmapLabels } from "./heatmap-labels";
+import { HeatMapRectangles } from "./heatmap-rectangles";
+import createHighlightClauses from "./utils/create-highlight-clauses";
+import getHighlightPosition from "./utils/get-highlight-position";
+import getHoverPosition, { HoverPosition } from "./utils/get-hover-position";
+import { modalTitle } from "./utils/modal-title";
+import { nestedDataset } from "./utils/nested-dataset";
+import { ColorScale, LinearScale } from "./utils/scales";
+
+interface LabelledHeatmapProps {
+  stage: Stage;
+  essence: Essence;
+  dataset: Datum[];
+  xScale: LinearScale;
+  yScale: LinearScale;
+  colorScale: ColorScale;
+
+  highlight: Highlight | null;
+  saveHighlight: Unary<List<FilterClause>, void>;
+  dropHighlight: Fn;
+  acceptHighlight: Fn;
+}
+
+interface LabelledHeatmapState {
+  scrollLeft: number;
+  scrollTop: number;
+  hoverPosition: HoverPosition | null;
+  leftLabelsWidth: number;
+  topLabelsHeight: number;
+}
+
+export const TILE_SIZE = 25;
+export const TILE_GAP = 2;
+const MIN_LEFT_LABELS_WIDTH = 100;
+const MAX_LEFT_LABELS_WIDTH = 200;
+const MIN_TOP_LABELS_HEIGHT = 100;
+const MAX_TOP_LABELS_HEIGHT = 150;
+
+function formatSegments(dataset: Datum[], fieldName: string, timezone: Timezone): string[] {
+  return dataset.map(datum => formatSegment(datum[fieldName], timezone));
+}
+
+export class LabelledHeatmap extends React.PureComponent<LabelledHeatmapProps, LabelledHeatmapState> {
+  state: LabelledHeatmapState = {
+    hoverPosition: null,
+    leftLabelsWidth: 0,
+    topLabelsHeight: 0,
+    scrollLeft: 0,
+    scrollTop: 0
+  };
+
+  saveHover = (x: number, y: number, part: ScrollerPart) => {
+    const { xScale, yScale } = this.props;
+    const hoverPosition = getHoverPosition(xScale, yScale, x, y, part, this.layout());
+    this.setState({ hoverPosition });
+  };
+
+  resetHover = () => this.setState({ hoverPosition: null });
+
+  saveScroll = (scrollTop: number, scrollLeft: number) => this.setState({ scrollLeft, scrollTop });
+
+  saveLeftLabelWidth = (maxLabelWidth: number) => this.setState({ leftLabelsWidth: Math.min(maxLabelWidth, MAX_LEFT_LABELS_WIDTH) });
+
+  saveTopLabelHeight = (maxLabelHeight: number) => this.setState({ topLabelsHeight: Math.min(maxLabelHeight, MAX_TOP_LABELS_HEIGHT) });
+
+  handleHighlight = (x: number, y: number, part: ScrollerPart) => {
+    if (part === "body" || part === "top-gutter" || part === "left-gutter") {
+      const { saveHighlight, essence, dataset } = this.props;
+      const layout = this.layout();
+      const clauses = createHighlightClauses({ x: x - layout.left, y: y - layout.top, part }, essence, dataset);
+      saveHighlight(List(clauses));
+    }
+  };
+
+  private layout(): ScrollerLayout {
```

Comment:
Some tests...?

Reply:
What cases? Three for each clamp? And one for rest of the props (no branching)?